README.md

# Colorswatches

Colorswatches was created circa 2005 when I was using [Plua](http://en.wikipedia.org/wiki/Plua) to develop software for
 [Palm OS](http://en.wikipedia.org/wiki/Palm_OS) devices. I resurrected it recently (summer 2014) because I needed something similar while developing the [blink library](http://github.com/profburke/FILL-THIS-IN) for Lua.

If I really wanted to beat a dead horse, there are a number of other color collections I could implement such as
the ones at [Rapid Tables](http://www.rapidtables.com/web/color/RGB_Color.htm), or the list of [Crayola colors](http://en.wikipedia.org/wiki/List_of_Crayola_crayon_colors).


## Usage

    colors = require 'colorswatches'
    c = colors.red -- or colors['red']
    print(c.name, c.r, c.g, c.b)
    print(c) -- make use of Color's __tostring
    print(c:rgb()) -- rgb() returns multiple values

    myred = colors.new('AwesomeRed', 240, 12, 14)

This project uses semantic versioning. See <a href="http://semver.org">semver.org</a> for more information.


## Requirements and Installation

The library was developed and tested using [Lua](http://lua.org) 5.2.3. However it should run on other versions of Lua with little or no modification.


If you have downloaded the github repository, you can install the library by copying the files in the `src` subdirectory to
your Lua library (`/usr/local/share/lua/5.2` or similar). Alternatively, you can install it via [luarocks](http://luarocks.org/)

    luarocks install colors


## Documentation and Contact Information

More documentation (than this dirt simple library requires) can be found in the `doc` subdirectory.

The best way to contact me regarding this library is to post an issue at the [github repository](https://github.com/profburke/colorswatches/issues).


## License

Colorswatches is free software distributed under the terms of the MIT license. It may be used for any purpose, including commercial purposes, at absolutely no cost without having to ask us. The only requirement is that if you do use Lblink, then you should give us credit by including the appropriate copyright notice somewhere in your product or its documentation. For details, see `LICENSE`.







CHANGELOG.md
## Changelog for Lua Colors library

### 2014-08-29 Version 0.9.1
	This is the first released version.

### 2014-08-29 Version 0.9.2
	Renamed to avoid class with existing module.
	I suppose technically I haven't renumbered this in proper accord
	with semantic versioning, but nobody but me is using thie library
	yet.

	


src/colors.lua
--- Provides a collection of basic color names and their RGB values.
-- Plus a function for creating new color objects.
--
-- Colors provides a collection of RGB values keyed by color name.
-- Specialized collections of color names are provided as submodules.
--
-- @module colors
-- @author Matthew M. Burke
-- @copyright 2005-2014
-- @license MIT (see LICENSE file)
--

local colors_mt = {}
local color_mt = {}
colors_mt.__index = colors_mt
color_mt.__index = color_mt


local colors = {}
setmetatable(colors, colors_mt)




--- Creates a new color object. This object
-- has fields *name*, *r*, *g*, and *b*.
-- @function new
-- @tparam string n name of the color
-- @tparam int r red value of the color (0-255)
-- @tparam int g green value of the color (0-255)
-- @tparam int b blue value of the color (0-255)
-- @treturn Color a table representing the color.
colors_mt.new = function(n, r, g, b)
   local color = { name = n, r = r, g = g, b = b }
   setmetatable(color, color_mt)
   return color
end




--- @type Color

--- Returns a string representation of the color including both its name and
-- its RGB values.
-- @function __tostring
-- @treturn string the string representation of the color.
color_mt.__tostring = function(s) 
   return string.format("color('%s', %d, %d, %d)", s.name, s.r, s.g, s.b)
end



--- Returns the RGB values of the color. They are returned as three separate values.
-- @function rgb
-- @treturn int red value (0-255)
-- @treturn int green (0-255)
-- @treturn int blue (0-255)
color_mt.rgb = function(s)
   return s.r, s.g, s.b
end




-- convenience method to define a color and store
-- it in the colors table
local color = function(n, r, g, b)
   local c = colors_mt.new(n, r, g, b)
   colors[c.name] = c
end




color('white', 255, 255, 255)
color('black', 0, 0, 0)
color('red', 255, 0, 0)
color('green', 0, 255, 0)
color('blue', 0, 0, 255)
color('yellow', 255, 255, 0)
color('cyan', 0, 255, 255)
color('magenta', 255, 0, 255)


return colors



src/colorswatches.lua
--- Provides a collection of basic color names and their RGB values.
-- Plus a function for creating new color objects.
--
-- Colorswatches provides a collection of RGB values keyed by color name.
-- Specialized collections of color names are provided as submodules.
--
-- @module colorswatches
-- @author Matthew M. Burke
-- @copyright 2005-2014
-- @license MIT (see LICENSE file)
--

local colors_mt = {}
local color_mt = {}
colors_mt.__index = colors_mt
color_mt.__index = color_mt


local colors = {}
setmetatable(colors, colors_mt)




--- Creates a new color object. This object
-- has fields *name*, *r*, *g*, and *b*.
-- @function new
-- @tparam string n name of the color
-- @tparam int r red value of the color (0-255)
-- @tparam int g green value of the color (0-255)
-- @tparam int b blue value of the color (0-255)
-- @treturn Color a table representing the color.
colors_mt.new = function(n, r, g, b)
   local color = { name = n, r = r, g = g, b = b }
   setmetatable(color, color_mt)
   return color
end




--- @type Color

--- Returns a string representation of the color including both its name and
-- its RGB values.
-- @function __tostring
-- @treturn string the string representation of the color.
color_mt.__tostring = function(s) 
   return string.format("color('%s', %d, %d, %d)", s.name, s.r, s.g, s.b)
end



--- Returns the RGB values of the color. They are returned as three separate values.
-- @function rgb
-- @treturn int red value (0-255)
-- @treturn int green (0-255)
-- @treturn int blue (0-255)
color_mt.rgb = function(s)
   return s.r, s.g, s.b
end




-- convenience method to define a color and store
-- it in the colors table
local color = function(n, r, g, b)
   local c = colors_mt.new(n, r, g, b)
   colors[c.name] = c
end




color('white', 255, 255, 255)
color('black', 0, 0, 0)
color('red', 255, 0, 0)
color('green', 0, 255, 0)
color('blue', 0, 0, 255)
color('yellow', 255, 255, 0)
color('cyan', 0, 255, 255)
color('magenta', 255, 0, 255)


return colors



src/colorswatches/x11.lua
--- Provides a collection of X11 color names and their RGB values.
-- For more details, see the documentation for `colorswatches`.
--
-- @module colorswatches.x11
--
-- @usage
-- x11_colors = require 'colorswatches.x11'
--
--
-- -- print out the names and RGB values
-- for name, color in pairs(x11_colors) do
--       print(name, color:rgb())
-- end
--
--
-- -- or get the components individually
-- crimson = x11_colors.crimson
-- print(crimson.name, crimson.r, crimson.g, crimson.b)
--
--
-- -- or make use of color's __tostring method
-- print(crimson)
--


local colors = require 'colorswatches'
colors.x11 = {}
local x11_colors = colors.x11

local colors_mt = getmetatable(colors)
setmetatable(x11_colors, colors_mt)


local color = function(n, r, g, b)
   local c = colors.new(n, r, g, b)
   x11_colors[c.name] = c
end



color('white', 255, 255, 255)
color('black', 0, 0, 0)

color('red', 255, 0, 0)
color('green', 0, 255, 0)
color('blue', 0, 0, 255)

color('yellow', 255, 255, 0)
color('cyan', 0, 255, 255)
color('magenta', 255, 0, 255)

color('whitesmoke', 245, 245, 245) 
color('gainsboro', 220, 220, 220)
color('lightgrey', 211, 211, 211)
color('silver', 192, 192, 192)
color('lightslategray', 119, 136, 153)
color('steelblue', 70, 130, 180)
color('royalblue', 65, 105, 225)
color('cornflowerblue', 100, 149, 237)
color('lightsteelblue', 176, 196, 222)
color('mediumslateblue', 123, 104, 238)
color('slateblue', 106, 90, 205)
color('dodgerblue', 30, 144, 255)
color('deepskyblue', 0, 191, 255)
color('lightskyblue', 135, 206, 250)
color('skyblue', 135, 206, 235)
color('lightblue', 173, 216, 230)
color('powderblue', 176, 224, 230)
color('azure', 240, 255, 255)
color('lightcyan', 224, 255, 255)
color('paleturquoise', 175, 238, 238)
color('darkturquoise', 0, 206, 209)
color('cadetblue', 95, 158, 160)
color('mediumturquoise', 72, 209, 204)
color('aqua', 0, 255, 255)
color('turquoise', 64, 224, 208)
color('lightseagreen', 32, 178, 170)
color('darkcyan', 0, 139, 139)
color('aquamarine', 127, 255, 212)
color('mediumaquamarine', 102, 205, 170)
color('darkseagreen', 143, 188, 143)
color('mediumseagreen', 60, 179, 113)
color('seagreen', 46, 139, 87)
color('limegreen', 50, 205, 50)
color('lime', 0, 255, 0)
color('chartreuse', 127, 255, 0)
color('lawngreen', 124, 252, 0)
color('greenyellow', 173, 255, 47)
color('palegreen', 152, 251, 152)
color('lightgreen', 144, 238, 144)
color('springgreen', 0, 255, 127)
color('mediumspringgreen', 0, 250, 154)
color('olivedrab', 107, 142, 35)
color('darkkhaki', 189, 183, 107)
color('darkgoldenrod', 184, 134, 11)
color('goldenrod', 218, 165, 32)
color('gold', 255, 215, 0)
color('khaki', 240, 230, 140)
color('palegoldenrod', 238, 232, 170)
color('blanchedalmond', 255, 235, 205)
color('moccasin', 255, 228, 181)
color('wheat', 245, 222, 179)
color('navajowhite', 255, 222, 173)
color('burlywood', 222, 184, 135)
color('tan', 210, 180, 140)
color('rosybrown', 188, 143, 143)
color('sienna', 160, 82, 45)
color('saddlebrown', 139, 69, 19)
color('chocolate', 210, 105, 30)
color('peru', 205, 133, 63)
color('sandybrown', 244, 164, 96)
color('brown', 165, 42, 42)
color('indianred', 205, 92, 92)
color('lightcoral', 240, 128, 128)
color('salmon', 250, 128, 114)
color('darksalmon', 233, 150, 122)
color('lightsalmon', 255, 160, 122)
color('coral', 255, 127, 80)
color('tomato', 255, 99, 71)
color('darkorange', 255, 140, 0)
color('orange', 255, 165, 0)
color('orangered', 255, 69, 0)
color('crimson', 220, 20, 60)
color('deeppink', 255, 20, 147)
color('fuchsia', 255, 0, 255)
color('hotpink', 255, 105, 180)
color('lightpink', 255, 182, 193)
color('pink', 255, 192, 203)
color('palevioletred', 219, 112, 147)
color('mediumvioletred', 199, 21, 133)
color('mediumpurple', 147, 112, 219)
color('darkorchid', 153, 50, 204)
color('mediumorchid', 186, 85, 211)
color('orchid', 218, 112, 214)
color('violet', 238, 130, 238)
color('plum', 221, 160, 221)
color('thistle', 216, 191, 216)
color('lavender', 230, 230, 250)
color('ghostwhite', 248, 248, 255)
color('aliceblue', 240, 248, 255)
color('mintcream', 245, 255, 250)
color('honeydew', 240, 255, 240)
color('lightgoldenrodyellow', 250, 250, 210)
color('lemonchiffon', 255, 250, 205)
color('cornsilk', 255, 248, 220)
color('lightyellow', 255, 255, 224)
color('ivory', 255, 255, 240)
color('floralwhite', 255, 250, 240)
color('linen', 250, 240, 230)
color('oldlace', 253, 245, 230)
color('antiquewhite', 250, 235, 215)
color('bisque', 255, 228, 196)
color('peachpuff', 255, 218, 185)
color('papayawhip', 255, 239, 213)
color('seashell', 255, 245, 238)
color('lavenderblush', 255, 240, 245)
color('mistyrose', 255, 228, 225)



return x11_colors


src/colorswatches/css.lua
--- Provides a collection of CSS color names and their RGB values.
-- For more details, see the documentation for `colorswatches`.
--
-- @module colorswatches.css
--
-- @usage
-- css_colors = require 'colorswatches.css'
--
--
-- -- print out the names and RGB values
-- for name, color in pairs(css_colors) do
--       print(name, color:rgb())
-- end
--
--
-- -- or get the components individually
-- goldenrod = css_colors.goldenrod
-- print(goldenrod.name, goldenrod.r, goldenrod.g, goldenrod.b)
--
--
-- -- or make use of color's __tostring method
-- print(goldenrod)
--


local colors = require 'colorswatches'
colors.css = {}
local css_colors = colors.css

local colors_mt = getmetatable(colors)
setmetatable(css_colors, colors_mt)


local color = function(n, r, g, b)
   local c = colors.new(n, r, g, b)
   css_colors[c.name] = c
end


color('aliceblue', 240, 248, 255)
color('antiquewhite', 250, 235, 215)
color('aqua', 0, 255, 255)
color('aquamarine', 127, 255, 212)
color('azure', 240, 255, 255)
color('beige', 245, 245, 220)
color('bisque', 255, 228, 196)
color('black', 0, 0, 0)
color('blanchedalmond', 255, 235, 205)
color('blue', 0, 0, 255)
color('blueviolet', 138, 43, 226)
color('brown', 165, 42, 42)
color('burlywood', 222, 184, 135)
color('cadetblue', 95, 158, 160)
color('chartreuse', 127, 255, 0)
color('chocolate', 210, 105, 30)
color('coral', 255, 127, 80)
color('cornflowerblue', 100, 149, 237)
color('cornsilk', 255, 248, 220)
color('crimson', 220, 20, 60)
color('cyan', 0, 255, 255)
color('darkblue', 0, 0, 139)
color('darkcyan', 0, 139, 139)
color('darkgoldenrod', 184, 134, 11)
color('darkgray', 169, 169, 169)
color('darkgreen', 0, 100, 0)
color('darkgrey', 169, 169, 169)
color('darkkhaki', 189, 183, 107)
color('darkmagenta', 139, 0, 139)
color('darkolivegreen', 85, 107, 47)
color('darkorange', 255, 140, 0)
color('darkorchid', 153, 50, 204)
color('darkred', 139, 0, 0)
color('darksalmon', 233, 150, 122)
color('darkseagreen', 143, 188, 143)
color('darkslateblue', 72, 61, 139)
color('darkslategray', 47, 79, 79)
color('darkslategrey', 47, 79, 79)
color('darkturquoise', 0, 206, 209)
color('darkviolet', 148, 0, 211)
color('deeppink', 255, 20, 147)
color('deepskyblue', 0, 191, 255)
color('dimgray', 105, 105, 105)
color('dimgrey', 105, 105, 105)
color('dodgerblue', 30, 144, 255)
color('firebrick', 178, 34, 34)
color('floralwhite', 255, 250, 240)
color('forestgreen', 34, 139, 34)
color('fuchsia', 255, 0, 255)
color('gainsboro', 220, 220, 220)
color('ghostwhite', 248, 248, 255)
color('gold', 255, 215, 0)
color('goldenrod', 218, 165, 32)
color('gray', 128, 128, 128)
color('green', 0, 128, 0)
color('greenyellow', 173, 255, 47)
color('grey', 128, 128, 128)
color('honeydew', 240, 255, 240)
color('hotpink', 255, 105, 180)
color('indianred', 205, 92, 92)
color('indigo', 75, 0, 130)
color('ivory', 255, 255, 240)
color('khaki', 240, 230, 140)
color('lavender', 230, 230, 250)
color('lavenderblush', 255, 240, 245)
color('lawngreen', 124, 252, 0)
color('lemonchiffon', 255, 250, 205)
color('lightblue', 173, 216, 230)
color('lightcoral', 240, 128, 128)
color('lightcyan', 224, 255, 255)
color('lightgoldenrodyellow', 250, 250, 210)
color('lightgray', 211, 211, 211)
color('lightgreen', 144, 238, 144)
color('lightgrey', 211, 211, 211)
color('lightpink', 255, 182, 193)
color('lightsalmon', 255, 160, 122)
color('lightseagreen', 32, 178, 170)
color('lightskyblue', 135, 206, 250)
color('lightslategray', 119, 136, 153)
color('lightslategrey', 119, 136, 153)
color('lightsteelblue', 176, 196, 222)
color('lightyellow', 255, 255, 224)
color('lime', 0, 255, 0)
color('limegreen', 50, 205, 50)
color('linen', 250, 240, 230)
color('magenta', 255, 0, 255)
color('maroon', 128, 0, 0)
color('mediumaquamarine', 102, 205, 170)
color('mediumblue', 0, 0, 205)
color('mediumorchid', 186, 85, 211)
color('mediumpurple', 147, 112, 219)
color('mediumseagreen', 60, 179, 113)
color('mediumslateblue', 123, 104, 238)
color('mediumspringgreen', 0, 250, 154)
color('mediumturquoise', 72, 209, 204)
color('mediumvioletred', 199, 21, 133)
color('midnightblue', 25, 25, 112)
color('mintcream', 245, 255, 250)
color('mistyrose', 255, 228, 225)
color('moccasin', 255, 228, 181)
color('navajowhite', 255, 222, 173)
color('navy', 0, 0, 128)
color('oldlace', 253, 245, 230)
color('olive', 128, 128, 0)
color('olivedrab', 107, 142, 35)
color('orange', 255, 165, 0)
color('orangered', 255, 69, 0)
color('orchid', 218, 112, 214)
color('palegoldenrod', 238, 232, 170)
color('palegreen', 152, 251, 152)
color('paleturquoise', 175, 238, 238)
color('palevioletred', 219, 112, 147)
color('papayawhip', 255, 239, 213)
color('peachpuff', 255, 218, 185)
color('peru', 205, 133, 63)
color('pink', 255, 192, 203)
color('plum', 221, 160, 221)
color('powderblue', 176, 224, 230)
color('purple', 128, 0, 128)
color('red', 255, 0, 0)
color('rosybrown', 188, 143, 143)
color('royalblue', 65, 105, 225)
color('saddlebrown', 139, 69, 19)
color('salmon', 250, 128, 114)
color('sandybrown', 244, 164, 96)
color('seagreen', 46, 139, 87)
color('seashell', 255, 245, 238)
color('sienna', 160, 82, 45)
color('silver', 192, 192, 192)
color('skyblue', 135, 206, 235)
color('slateblue', 106, 90, 205)
color('slategray', 112, 128, 144)
color('slategrey', 112, 128, 144)
color('snow', 255, 250, 250)
color('springgreen', 0, 255, 127)
color('steelblue', 70, 130, 180)
color('tan', 210, 180, 140)
color('teal', 0, 128, 128)
color('thistle', 216, 191, 216)
color('tomato', 255, 99, 71)
color('turquoise', 64, 224, 208)
color('violet', 238, 130, 238)
color('wheat', 245, 222, 179)
color('white', 255, 255, 255)
color('whitesmoke', 245, 245, 245)
color('yellow', 255, 255, 0)
color('yellowgreen', 154, 205, 50)


return css_colors
